### PR TITLE
WORKFLOWS-21: Add ECS cluster stack

### PR DESCRIPTION
**Pull request by tthyer**
[WORKFLOWS-21](https://sagebionetworks.jira.com/browse/WORKFLOWS-21)
This creates an ECS cluster that has a single instance in it.

One disadvantage of doing this is that we can't export the PublicIp attribute as we would if this were configured as `AWS::EC2::Instance`, and that is required for the [ECS task stack](https://sagebionetworks.jira.com/browse/WORKFLOWS-22). However, I will probably be able to use `!rcmd` to get the PublicIP for that parameter.

PR depends on #19 and on https://github.com/Sage-Bionetworks/aws-infra/pull/315

### CODE REVIEW

File: config/develop/nextflow-ecs-capacity.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+template_path: nextflow-ecs-capacity.yaml
+stack_name: nextflow-ecs-capacity
+parameters:
+  EcsClusterName: !stack_output_external nextflow-ecs-cluster::EcsClusterName
+  AutoScalingGroupArn: !rcmd >-
+    aws autoscaling describe-auto-scaling-groups
+    | jq -r '.AutoScalingGroups | .[] | select(.AutoScalingGroupName | startswith("nextflow-ecs-cluster-EcsAutoScalingGroup")) | .AutoScalingGroupARN'
```
Comment (tthyer) [inline config/develop/nextflow-ecs-capacity.yaml:7]: This unusual way of deriving this parameter arises from the lack of ARN exposure for AutoScalingGroup -- it is not the Ref nor accessible through GetAtt.

File: Pipfile [hunk 2]
```diff
@@ -4,10 +4,11 @@ verify_ssl = true
 name = "pypi"
 
 [packages]
-sceptre = "2.4.0"
```
Comment (tthyer) [inline Pipfile:7]: Please ignore changes to Pipfile and Pipfile.lock -- these changes are being made separately in #19 .

**Review by thomasyu888 — APPROVED** [review]
LGTM

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -9,7 +9,7 @@ repos:
     hooks:
       - id: yamllint
   - repo: https://github.com/awslabs/cfn-python-lint
-    rev: v0.29.0
+    rev: v0.52.0
```
Comment (tthyer) [inline .pre-commit-config.yaml:12]: There's another workflows issue to update all the pre-commit software, but this one had to be updated now to recognize new Cloudformation types.

File: config/develop/nextflow-ecs-capacity.yaml (hunk 1 above)
Comment (BrunoGrandePhD) [reply config/develop/nextflow-ecs-capacity.yaml:9]: Should this be a comment in the config for future readers like Paolo?

**Review by BrunoGrandePhD — APPROVED** [review]
**Full Disclosure:** I'm not well-versed enough in CloudFormation to notice subtle issues. Otherwise, this looks good to me!

File: templates/nextflow-ecs-capacity.yaml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      Cluster: !Ref EcsClusterName
+      DefaultCapacityProviderStrategy:
+        - CapacityProvider: !Ref EcsCapacityProvider
+
+Outputs:
+
+  EcsCapacityProviderName:
+    Value: !Ref EcsCapacityProvider
+    Export:
+      Name: !Sub '${AWS::Region}-${AWS::StackName}-EcsCapacityProvider'
```
Comment (zaro0508) [inline templates/nextflow-ecs-capacity.yaml:39]: EcsCapacityProvider/EcsCapacityProviderName ?

File: templates/nextflow-ecs-cluster.yaml [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+          FromPort: !Ref SecurityIngressFromPort
+          ToPort: !Ref SecurityIngressToPort
+          CidrIp: !Ref SecurityIngressCidrIp
+
+  EcsLaunchConfiguration:
+    Type: AWS::AutoScaling::LaunchConfiguration
+    Properties:
+      ImageId: !Ref EcsAmiId
+      InstanceType: !Ref EcsInstanceType
+      AssociatePublicIpAddress: true
```
Comment (zaro0508) [inline templates/nextflow-ecs-cluster.yaml:135]: hmm, i see you've passed in private subnets for `Subnet` however your associating public IPs?  this seems contradictory?

all instance are public?  don't we prefer them to be private for security reasons?

File: templates/nextflow-ecs-cluster.yaml [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+      IamInstanceProfile: !Ref EcsInstanceProfile
+      SecurityGroups:
+        - !Ref EcsSecurityGroup
```
Comment (zaro0508) [inline templates/nextflow-ecs-cluster.yaml:138]: opening up ports to public instances is usually not recommened.  Is there putting instances in the public?

**Review by zaro0508 — CHANGES_REQUESTED** [review]
Have you thought about using Fargate with ECS?  I've heard that it can make things much easier.

File: config/develop/nextflow-ecs-capacity.yaml [hunk 7]
```diff
@@ -0,0 +1,13 @@
+template_path: nextflow-ecs-capacity.yaml
+stack_name: nextflow-ecs-capacity
+dependencies:
+  - develop/nextflow-ecs-cluster.yaml
+parameters:
+  EcsClusterName: !stack_output_external nextflow-ecs-cluster::EcsClusterName
+  AutoScalingGroupArn: !rcmd >-
```
[the remark this answers is not in the document]
Comment (zaro0508) [reply config/develop/nextflow-ecs-capacity.yaml:7]: You might be interested in the following PRs..
https://github.com/Sceptre/sceptre-resolver-cmd/pull/5
https://github.com/Sceptre/sceptre-resolver-cmd/pull/6

[the remark this answers is not in the document]
Comment (tthyer) [reply config/develop/nextflow-ecs-capacity.yaml:7]: `aws-actions/configure-aws-credentials@v1` sets the environment variables. We don't have to pass in a profile.

File: templates/nextflow-ecs-cluster.yaml (hunk 5 above)
Comment (tthyer) [reply templates/nextflow-ecs-cluster.yaml:135]: Good catch -- this cluster is specifically for the NF tower software. We need to be able to access its frontend in a browser, so I'll change these to a single public subnet, unless you have a better solution.
All of the other clusters are created through NextFlow forge for running data jobs, and those will use exclusively private subnets.

File: templates/nextflow-ecs-cluster.yaml (hunk 6 above)
Comment (tthyer) [reply templates/nextflow-ecs-cluster.yaml:138]: see above

File: config/develop/nextflow-ecs-capacity.yaml (hunk 1 above)
Comment (tthyer) [reply config/develop/nextflow-ecs-capacity.yaml:11]: Good suggestion -- added.

File: .pre-commit-config.yaml (hunk 3 above)
Comment (tthyer) [reply .pre-commit-config.yaml:12]: did that, but just the once to make the updates, not in the build file

File: templates/nextflow-ecs-capacity.yaml (hunk 4 above)
Comment (tthyer) [reply templates/nextflow-ecs-capacity.yaml]: changed name

File: templates/nextflow-ecs-cluster.yaml (hunk 5 above)
Comment (tthyer) [reply templates/nextflow-ecs-cluster.yaml:135]: @zaro0508 Since you're out this week I am going to follow up with you in email

**Comment by tthyer** [comment]
@zaro0508 Yes, have experience with Fargate. It's pretty neat but has a low resource bar that makes it bad for some bioinformatics applications. In this case I'm fulfilling a specification but Fargate might work. But not going to explore that atm -- if I were going to rework this infra would rewrite with EKS instead.

**Pull request merged**